africanus/dft/kernels.py

```diff
@@ -44,21 +44,19 @@ def _im_to_vis_impl(image, uvw, lm, frequency, dtype=None):
                 for chan in range(frequency.shape[0]):
                     p = real_phase * frequency[chan] * 1.0j
 
-                    # Our phase input is purely imaginary
-                    # so we can can elide a call to exp
-                    # and just compute the cos and sin
-                    # @simon does this really make a difference?
-                    # I thought a complex exponential is evaluated
-                    # as a sum of sin and cos anyway
-                    vis_of_im[row, chan] += np.exp(p)*image[source, chan]
+                    for corr in range(image.shape[-1]):
+                        if image[source, chan, corr]:
+                            vis_of_im[row, chan,
+                                      corr] += np.exp(p)*image[source,
+                                                               chan, corr]
```

Comment:
Perhaps shorten the loop vars to from `(row, source, chan, corr)` to `(r, s, f, c)`:

```python
for r in range(nrows):
  for s in range(nsrc):
     for f in range(nchan):
       for c in range(ncorr):
         vis_of_im[r, f, c] += np.exp(p)*image[s, f, c]